presto-main/src/test/java/io/prestosql/sql/planner/assertions/IndexJoinMatcher.java

```diff
@@ -0,0 +1,103 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.assertions;
+
+import com.google.common.collect.ImmutableSet;
+import io.prestosql.Session;
+import io.prestosql.cost.StatsProvider;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.sql.planner.plan.IndexJoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.prestosql.sql.planner.assertions.MatchResult.NO_MATCH;
+import static java.util.Objects.requireNonNull;
+
+final class IndexJoinMatcher
+        implements Matcher
+{
+    private final IndexJoinNode.Type type;
+    private final List<ExpectedValueProvider<IndexJoinNode.EquiJoinClause>> criteria;
+    private final Optional<PlanTestSymbol> probeHashSymbol;
+    private final Optional<PlanTestSymbol> indexHashSymbol;
+
+    IndexJoinMatcher(
+            IndexJoinNode.Type type,
+            List<ExpectedValueProvider<IndexJoinNode.EquiJoinClause>> criteria,
+            Optional<PlanTestSymbol> probeHashSymbol,
+            Optional<PlanTestSymbol> indexHashSymbol)
+    {
+        this.type = requireNonNull(type, "type is null");
+        this.criteria = requireNonNull(criteria, "criteria is null");
+        this.probeHashSymbol = requireNonNull(probeHashSymbol, "probeHashSymbol is null");
+        this.indexHashSymbol = requireNonNull(indexHashSymbol, "indexHashSymbol is null");
+    }
+
+    @Override
+    public boolean shapeMatches(PlanNode node)
+    {
+        if (!(node instanceof IndexJoinNode)) {
+            return false;
+        }
+
+        IndexJoinNode indexJoinNode = (IndexJoinNode) node;
+        return indexJoinNode.getType() == type;
+    }
+
+    @Override
+    public MatchResult detailMatches(PlanNode node, StatsProvider stats, Session session, Metadata metadata, SymbolAliases symbolAliases)
+    {
+        checkState(shapeMatches(node), "Plan testing framework error: shapeMatches returned false in detailMatches in %s", this.getClass().getName());
+        IndexJoinNode indexJoinNode = (IndexJoinNode) node;
+
+        if (indexJoinNode.getCriteria().size() != criteria.size()) {
+            return NO_MATCH;
+        }
+        Set<IndexJoinNode.EquiJoinClause> actualCriteria = ImmutableSet.copyOf(indexJoinNode.getCriteria());
+        Set<IndexJoinNode.EquiJoinClause> expectedCriteria = criteria.stream()
+                .map(maker -> maker.getExpectedValue(symbolAliases))
```

Comment:
What does "maker" mean for the variable name? Was it supposed to be "marker" (i.e., a typo)?